src/others/contexts/dictionary-context.tsx

```diff
@@ -0,0 +1,79 @@
+import { ReactNode, useState, useContext, createContext, useEffect } from "react";
+import { useLocationsQuery, useSuppliesQuery, Supply, Location } from "./api";
+
+interface DictionaryState {
+  suppliesDict: SuppliesDict | undefined; // TODO: it's used only in map - doesn't have to be exposed once map is using the declarative API (translate...)
+  locationDict: LocationsDict | undefined; // TODO: it's used only in map - doesn't have to be exposed once map is using the declarative API (translate...)
+  translateLocation: (city_id: number) => Location | undefined;
+  translateSupply: (category_id: string) => Supply | undefined;
+}
+
+type LocationsDict = { [k: string]: Location };
+type SuppliesDict = { [k: string]: Supply };
+
+export const DictionaryContext = createContext<DictionaryState | null>(null);
+
+export const useDictionaryContext = () => {
+  const dictionaryContext = useContext(DictionaryContext);
+  if (!dictionaryContext) {
+    throw new Error("dictionaryContext must be used within dictionaryContextProvider");
+  }
+  return dictionaryContext;
+};
+
+interface DictionaryContextProviderProps {
+  children: ReactNode;
+}
+
+export const DictionaryContextProvider = ({ children }: DictionaryContextProviderProps) => {
+  const { data: locations } = useLocationsQuery();
+  const { data: supplies } = useSuppliesQuery();
+
+  const dictionaryState = useDictionaryState({ supplies, locations });
+
+  return <DictionaryContext.Provider value={dictionaryState}>{children}</DictionaryContext.Provider>;
+};
+
+const useDictionaryState = ({ supplies, locations }: { supplies: Supply[] | undefined; locations: Location[] | undefined }) => {
+  const [suppliesDict, setSuppliesDict] = useState<SuppliesDict | undefined>(undefined);
+  const [locationDict, setLocationsDict] = useState<LocationsDict | undefined>(undefined);
+
+  useEffect(() => {
+    if (locations?.length) {
+      initLocationsDict(locations);
+    }
+
+    if (supplies?.length) {
+      initSuppliesDict(supplies);
+    }
+  }, [supplies, locations]);
+
+  const initSuppliesDict = (supplies: Supply[]) => {
+    const suppliesDict = supplies.reduce((acc, cur) => ({ ...acc, [cur.id]: cur }), {});
+    setSuppliesDict(suppliesDict);
+  };
+
+  const initLocationsDict = (locations: Location[]) => {
+    const locationsDict = locations.reduce((acc, cur) => ({ ...acc, [cur.id]: cur }), {});
```

Comment:
again maybe just my TS ignorance coming up here, but the way i read this line, it will run for
O(N^2) where N is the number of cities.
The reason I think it is O(N^2) is that you copy the whole dict on each addition ({...acc})

Given the 2k cities and slow JS, this might be significant.
Why not build a dict with just a plain loop in a linear time?

Reply:
I don't think JS is "slow". "Slow" compared to what? 

I also don't think it matters. If we have arrays with millions of items, it may, but for now I would not be that much concerned with optimalisation.

I believe in most of the cases, it's not JS inefficiency that makes UI slow. And loops are terrible to read, that's the general frontend people's opinion, I guess.

Reply:
I appreciate the comments though - make me think of things I'd never think otherwise.